[arrow] Update to 6.0.0

ianmcook opened the pull request:
Updates the arrow port to version 6.0.0

ianmcook committed:
Bump version to 6.0.0

ianmcook committed:
Update patch

ianmcook committed:
Update version files

JonLiu1993 commented:
@ianmcook ,Could you please take a look:
https://dev.azure.com/vcpkg/public/_build/results?buildId=62270&view=artifacts&pathAsName=false&type=publishedArtifacts

ianmcook committed:
Update patch

ianmcook committed:
Update version files

ianmcook committed:
Update patch

ianmcook committed:
Update version files

ianmcook committed:
Fix deprecated functions

ianmcook committed:
Update version files

ianmcook committed:
Fix unused variable warning

ianmcook committed:
Update version files

ianmcook commented:
@GPSnoopy here's the Arrow 6.0.0 update PR. Let me know if you encounter any issues when trying this with ParquetSharp.

JonLiu1993 commented:
@ianmcook ,Thanks for your contribution, Are you tested the features locally?

ianmcook committed:
Enclose paths in quotes

Co-authored-by: JonLiu1993 <63675417+JonLiu1993@users.noreply.github.com>

ianmcook committed:
Enclose paths in quotes

ianmcook committed:
Update version files

ianmcook committed:
Fix S3 feature

ianmcook committed:
Update version files

ianmcook committed:
Remove quotes from empty DZSTD_MSVC_LIB_PREFIX

ianmcook committed:
Update version files

ianmcook commented:
> Are you tested the features locally?

@JonLiu1993 I installed the following successfully in local environments:
```
arrow[core,csv,dataset,filesystem,json,mimalloc,orc,parquet,s3]:x64-windows-static
arrow[core,csv,dataset,filesystem,flight,jemalloc,json,mimalloc,orc,parquet,s3]:x64-osx
```
There are some issues with `s3` on `x64-windows` that I am troubleshooting.

The `flight` feature is experimental and does not build on Windows.

ianmcook commented:
The issue with `s3` on `x64-windows` is caused by these lines in `AWSSDKConfig.cmake`:
https://github.com/aws/aws-sdk-cpp/blob/2072c7cb48679d870aea3a3f6224306ba5ee66ef/cmake/AWSSDKConfig.cmake#L58-L62
With those lines removed, it installs successfully. But since it's not possible to remove them, how can we stop them from running when `find_package(AWSSDK)` is called in `ThirdpartyToolchain.cmake`? Setting `AWSSDK_INSTALL_AS_SHARED_LIBS` to `OFF` there (via a patch) has no effect; it is overridden and ends up being `ON` in `AWSSDKConfig.cmake`. The same happens for other variables. Any ideas for how to solve this? @kou please let me know if you have any suggestions, thanks.

jgiannuzzi commented:
@ianmcook ParquetSharp seems happy with the upgrade to Arrow 6.0.0 so far - here is the draft PR: https://github.com/G-Research/ParquetSharp/pull/230

I had to add `/Zc:__cplusplus` to make it work on Windows, but this might be due to our heavily-customised cmake setup.

kou commented:
@ianmcook Could you show full build log?

ianmcook commented:
@kou Here's the full log: https://gist.github.com/ianmcook/ff469c4d84e822ff5037c7072f230b68

kou commented:
@ianmcook Thanks. Could you try with `-DCMAKE_FIND_DEBUG_MODE=ON` to show more information?

ianmcook commented:
@kou I updated the gist to include a second file containing the tail of the log generated with `-DCMAKE_FIND_DEBUG_MODE=ON`. I didn't include the whole log because it's ~750K lines. I believe this shows the relevant part but let me know if you want to see more. Thanks.
https://gist.github.com/ianmcook/ff469c4d84e822ff5037c7072f230b68

kou commented:
@ianmcook Thanks. Could you also provide file list under ` C:/vcpkg/installed/x64-windows`?

ianmcook commented:
@kou Yes. I added `tree` output to the Gist. Thanks https://gist.github.com/ianmcook/ff469c4d84e822ff5037c7072f230b68#file-tree-out

kou commented:
@ianmcook Thanks. I think that we should report this to https://github.com/aws/aws-sdk-cpp and fix this in upstream instead of patching to Apache Arrow port.

FYI, MSYS2 package also disables the `set`: https://github.com/msys2/MINGW-packages/blob/master/mingw-w64-aws-sdk-cpp/0001-fix-dll-import-libary-folder.patch#L19

JonLiu1993 commented:
@ianmcook, thanks for your PR, when I tested the features in triple x64-windows on my local machine, it failed:
[config-x64-windows-out.log](https://github.com/microsoft/vcpkg/files/7552316/config-x64-windows-out.log)

dg0yt commented:
> thanks for your PR, when I tested the features in triple x64-windows on my local machine, it failed: [config-x64-windows-out.log](https://github.com/microsoft/vcpkg/files/7552316/config-x64-windows-out.log)

Oh, another case of "aws-sdk-cpp is unusable" in vcpkg...

dg0yt commented:
> I think that we should report this to https://github.com/aws/aws-sdk-cpp and fix this in upstream instead of patching to Apache Arrow port.

aws-sdk-cpp in vcpkg needs to be updated first.

dg0yt commented:
> The issue with `s3` on `x64-windows` is caused by these lines in `AWSSDKConfig.cmake`: https://github.com/aws/aws-sdk-cpp/blob/2072c7cb48679d870aea3a3f6224306ba5ee66ef/cmake/AWSSDKConfig.cmake#L58-L62 That causes `find_library()` later in the file to look in the wrong place (`bin` instead of `lib`) and throw the error
> 
> ```
> AWS SDK for C++ headers found, but we were unable to locate the binaries
> ```
> 
> With those lines removed, the install is successful. But since it's not possible to remove those lines directly, how can we stop them from running when `find_package(AWSSDK)` is called in `ThirdpartyToolchain.cmake`? Setting `AWSSDK_INSTALL_AS_SHARED_LIBS` to `OFF` there (via a patch) has no effect; it is overridden and ends up being `ON` in `AWSSDKConfig.cmake`. The same happens for other variables. Any ideas for how to solve this?

https://github.com/microsoft/vcpkg/pull/20289 removes that line.

dg0yt commented:
I expect that port aws-sdk-cpp needs more work to properly work with the vcpkg debug+release layout, in order to avoid mixing release and debug binaries.

jgiannuzzi commented:
FYI @ianmcook I have created a PR to enable arm64 builds of Arrow: https://github.com/microsoft/vcpkg/pull/21467
This is to enable us to build ParquetSharp on arm64 - see https://github.com/G-Research/ParquetSharp/issues/228

I'm happy to help rebasing one on top of the other, depending on which one gets merged first. The actual changes are fairly trivial:
* removed `vcpkg_fail_port_install`: https://github.com/microsoft/vcpkg/pull/21467/files#diff-363df8fbbe7343b382e186887095c573c6a0e78120eba6cf6c1941747f0898f7L1-L2
* updated `"supports"`: https://github.com/microsoft/vcpkg/pull/21467/files#diff-73abafa079174f6c3bf130b06465ec137f1789441af1cb3516653057d0d10e06L7-R7

Sorry for the potential extra work!

ianmcook commented:
Thanks @jgiannuzzi. No problem—if your PR merges first, I'll rebase this on it. Either way should be fine.

Thanks @JonLiu1993 @dg0yt for the details on `aws-sdk-cpp`. I guess let's hold off on merging this until #20289 is merged. After that I'll re-test features locally on `x64-windows` and reply here.

jgiannuzzi commented:
@ianmcook PR #21467 has been merged — it's rebase time! 😅

JonLiu1993 commented:
@ianmcook ,For response

ianmcook committed:
Update version files

ianmcook committed:
Update version files

ianmcook committed:
Update version files

ianmcook committed:
Update version files

ianmcook commented:
@jgiannuzzi Thanks! I rebased to pull in #21467.

@JonLiu1993 I rebased to pull in #20289 and re-tested features locally. `arrow[core,csv,dataset,filesystem,json,mimalloc,orc,parquet,s3]:x64-windows` now succeeds!

I think this is ready to merge.

BillyONeal commented on ports/arrow/portfile.cmake:
```diff
@@ -59,12 +59,16 @@ else()
     set(THRIFT_USE_SHARED ${ARROW_DEPENDENCY_USE_SHARED})
 endif()
 
+if(ARROW_S3)
```
Hmmm isn't this always false? `vcpkg_check_features` should only be setting this inside `FEATURE_OPTIONS`, not generally.

ianmcook replied:
Thanks for catching this; this was done in an unsuccessful attempt to work around the AWS SDK issue mentioned in earlier comments here. I think it's a no-op. I will remove it.

BillyONeal commented on ports/arrow/portfile.cmake:
```diff
@@ -59,12 +59,16 @@ else()
 endif()
 
+if(ARROW_S3)
+    set(S3_OPTIONS -DAWSSDK_ROOT_DIR="${CURRENT_PACKAGES_DIR}" -DAWSSDK_INSTALL_AS_SHARED_LIBS=OFF)
```
This is a little strange... why is it reading from `CURRENT_PACKAGES_DIR`? Note that that directory is likely to be gone when this would be useful?

BillyONeal commented on ports/arrow/fix-dependencies.patch:
```diff
@@ -67,7 +67,26 @@ index e6852d9c2..fc0973ffa 100644
+ # so we first need to determine whether we're building it
+ if(ARROW_WITH_THRIFT AND Thrift_SOURCE STREQUAL "AUTO")
+   find_package(Thrift 0.11.0 MODULE COMPONENTS ${ARROW_THRIFT_REQUIRED_COMPONENTS})
+-  if(Thrift_FOUND)
```
Can we set `THRIFT_SOURCE` to `BUNDLED` rather than patching this?

Will `BUNDLED` end up trying to use a vendored copy of thrift that might conflict with what other vcpkg ports are using?

ianmcook replied:
>Will BUNDLED end up trying to use a vendored copy of thrift that might conflict with what other vcpkg ports are using?

Yes, that is the concern. We need to do some testing to investigate this further before making the change. I opened #21821 for the arrow port maintainers to investigate this. If it's OK with the vcpkg maintainers, we would prefer to address this after this PR is merged.

ianmcook committed:
Remove old attempt to fix AWS SDK issue

ianmcook committed:
Update versions files

JonLiu1993 commented:
@BillyONeal, Could you please take a look

BillyONeal commented:
Thanks for the update and addressing the feedback; sorry it took so long to see that you fixed the main issue :)
